### PR TITLE
Parallelized download cell-health cell painting datasets from FigShare

### PR DESCRIPTION
**Pull request by axiomcura**
mp_datastream

This PR introduces the start of a new method to process cell-health cell painting data sets.

The primary focus of this PR is to imrpove the downloading performance of all cell-health cell painting data sets from Figshare data repository.  

The older version of `download.py` (source) streamed downloaded data into the data file rather than downloading the whole sqlite file into memory. This was advantageous as it can easily be parallelized.

The parallelized data streaming works by reserving X amount of memory to each CPU core. For example, the newer version of download.py has a cap of 750MB of reserved memory for data streaming. If  4 cores were used, then the downloading process will use a total of 3GB of memory (4cpu * 750MB).

## Performance check

The test conducted used 9 cores with a cap of 750MB reserved memory for streaming and below are the results. 

```
Now downloading... data/SQ00014612.sqlite
Now downloading... data/SQ00014610.sqlite
Now downloading... data/SQ00014611.sqlite
Now downloading... data/SQ00014613.sqlite
Now downloading... data/SQ00014614.sqlite
Now downloading... data/SQ00014616.sqlite
Now downloading... data/SQ00014615.sqlite
Now downloading... data/SQ00014617.sqlite
Now downloading... data/SQ00014618.sqlite

real    74m8.751s
user    6m29.535s
sys     12m37.347s
```

It took 75 minutes to download all 9 `sqlite` files. 

**Something to note about:** 

- Since we are using multiple cores to download into the hard drive, we could potentially be saturating the I/O transfer buffer (communication between CPU and Hard Drive) causing  `download.py`'s performance not to improve if more cores were added. 
- The performance is also completely dependent on the FigShare's uploading speed and your local internet speed. 

## aditional notes

- `download-pipeline.sh` contains commands to use both the `download.py` and `md5sum`
- md5sum is to check the quality of the downloaded dataset

### CODE REVIEW

**Comment by axiomcura** [comment]
@gwaybio  Here is the PR containing multi-process download

File: other_processing/0.download-profiles-from-figshare/download.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+"""
+Download Single Cell Cell Painting Profiles
+Gregory Way, 2019
```
Comment (gwaybio) [inline other_processing/0.download-profiles-from-figshare/download.py]: Please add yourself!

Maybe something like:

```python
"""
Download Single Cell Cell Painting Profiles in parallel
Erik Serrano, 2022
Modified from a script written by Gregory Way, 2019
https://github.com/broadinstitute/cell-health/blob/master/0.download-data/download.py
...
"""
```

File: other_processing/0.download-profiles-from-figshare/download.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+We then aggregated the output of CellProfiler using cytominer-database.
+Cytominer-database compiles all single cell profiles into a single `sqlite` database.
+We uploaded the `sqlite` files to NIH Figshare 
+https://nih.figshare.com/articles/dataset/Cell_Health_-_Cell_Painting_Single_Cell_Profiles/9995672/1.
+These data are publicly available
+
+*Important Note*
+
+These files are large (~130 GB Total).
+There is no need to download and reprocess the data.
```
Comment (gwaybio) [inline other_processing/0.download-profiles-from-figshare/download.py]: can you clarify if this is still the case?

Comment (gwaybio) [reply other_processing/0.download-profiles-from-figshare/download.py]: also, please add details about _why_ we're downloading.

It is to convert to `parquet` files and create version 6 of the data, correct?

**Review by gwaybio — APPROVED** [review]
LGTM

**Comment by gwaybio** [comment]
We also should keep your mermaid diagram somewhere as well!

Maybe in the `other_processing` README describing the approach?

File: other_processing/0.download-profiles-from-figshare/download.py (hunk 2 above)
Comment (axiomcura) [reply other_processing/0.download-profiles-from-figshare/download.py]: > can you clarify if this is still the case?
In terms of downloading, yes. Pipeline requires to download the raw dataset, which is 130GB. 

changes are applied on: 4ada9d2258d8f412b369ccdf4c47c9212278eff5

**Comment by axiomcura** [comment]
@gwaybio 
I updated the documentation in 4ada9d2258d8f412b369ccdf4c47c9212278eff5. 

Also, I am unable to merge!

**Comment by gwaybio** [comment]
@axiomcura - Thanks for letting me know. Would it make sense to add your mermaid diagram somewhere? If not, then I will merge now!

**Comment by axiomcura** [comment]
@gwaybio I'm currently creating another [branch](https://github.com/axiomcura/cell-health-data/tree/readme/other_processing) that is focused on creating the documentation for the new data version. The diagram is stored here!

**Comment by gwaybio** [comment]
Sounds good!

**Pull request merged**